File: packages/cli-py/planforge/utils/tui.py
"""Shared TUI key handling for arrow-key selection (doctor, model commands)."""

import sys

import readchar


def _normalize_key(key: str) -> str | None:
    """Map readchar key or escape sequence to up/down/left/right/enter/quit."""
    if key in ("\r", "\n"):
        return "enter"
    key_mod = getattr(readchar, "key", None)
    up = getattr(key_mod, "UP", None) if key_mod else None
    down = getattr(key_mod, "DOWN", None) if key_mod else None
    left = getattr(key_mod, "LEFT", None) if key_mod else None
    right = getattr(key_mod, "RIGHT", None) if key_mod else None
    if key in ("\x1b[A", "w", "k") or key == up:
        return "up"
    if key in ("\x1b[B", "s", "j") or key == down:
        return "down"
    if key in ("\x1b[D", "a", "h") or key == left:
        return "left"
    if key in ("\x1b[C", "d", "l") or key == right:
        return "right"
    if key == "\x03":  # Ctrl+C
        return "quit"
    return None


def wait_key() -> str | None:
    """Read one key and return up/down/left/right/enter/quit or None."""
    try:
        key = readchar.readkey()
    except KeyboardInterrupt:
        return "quit"
    return _normalize_key(key)


def select_from_list(
    items: list[tuple[str, object]],
    prompt: str,
    *,
    quit_label: str = "Quit",
    initial_index: int = 0,
) -> object | None:
    """Show list with arrow keys; Enter selects, Quit row or key returns None. items = [(label, value), ...]."""
    if not items:
        return None
    total_rows = len(items) + 1
    index = max(0, min(initial_index, len(items) - 1))
    print(f"\n  {prompt}\n")
    while True:
        for i, (label, _) in enumerate(items):
            prefix = "  > " if i == index else "    "
            print(f"{prefix}{label}")
        prefix = "  > " if index == len(items) else "    "
        print(f"{prefix}{quit_label}")
        key = wait_key()
        if key == "quit":
            return None
        if key == "enter":
            if index == len(items):
                return None
            return items[index][1]
        if key == "up":
            index = (index - 1) % total_rows
        elif key == "down":
            index = (index + 1) % total_rows
        sys.stdout.write(f"\033[{total_rows}A\033[0J")
        sys.stdout.flush()


def _format_role_line(role: str, role_config: dict) -> str:
    """Format a single role line for Current AI config display."""
    provider = role_config.get("provider", "")
    model = role_config.get("model", "")
    extra = ""
    if role_config.get("effort") is not None:
        extra = f" (effort: {role_config['effort']})"
    elif role_config.get("reasoning") is not None:
        extra = f" (reasoning: {role_config['reasoning']})"
    return f"  {role.ljust(12)}: {provider.ljust(6)} / {model.ljust(20)}{extra}"


def print_current_ai_config(
    config: dict,
    heading: str = "Current AI config",
) -> None:
    """Print planner/implementer config block (e.g. at start of planforge model)."""
    print(f"\n  {heading}")
    print(f"  {'-' * len(heading)}")
    print(_format_role_line("planner", config.get("planner", {})))
    print(_format_role_line("implementer", config.get("implementer", {})))

File: packages/cli-js/src/utils/tui.ts
/**
 * Shared TUI key handling for arrow-key selection (doctor, model commands).
 */

import type { PlanForgeConfig } from "../config/types.js";

export type KeyAction = "up" | "down" | "left" | "right" | "enter" | "quit" | null;

export interface SelectListItem<T> {
  label: string;
  value: T;
}

export function waitKey(): Promise<KeyAction> {
  return new Promise((resolveKey) => {
    if (!process.stdin.isTTY) {
      resolveKey(null);
      return;
    }
    const wasRaw = process.stdin.isRaw;
    process.stdin.setRawMode?.(true);
    process.stdin.resume();
    process.stdin.setEncoding("utf8");
    let buf = "";

    const resolveAndClean = (action: KeyAction) => {
      cleanup();
      resolveKey(action);
    };

    const onData = (chunk: string | Buffer) => {
      buf += typeof chunk === "string" ? chunk : chunk.toString("utf8");
      const s = buf;

      if (s === "\r" || s === "\n") {
        buf = "";
        resolveAndClean("enter");
        return;
      }
      if (s === "\u0003") {
        buf = "";
        resolveAndClean("quit");
        return;
      }
      // Arrow keys: full sequence (e.g. \x1b[A) or split (\x1b then [ then A)
      if (s.startsWith("\x1b[") && s.length >= 3) {
        const c = s[2];
        buf = s.length > 3 ? s.slice(3) : "";
        if (c === "A") resolveAndClean("up");
        else if (c === "B") resolveAndClean("down");
        else if (c === "C") resolveAndClean("right");
        else if (c === "D") resolveAndClean("left");
        else resolveAndClean(null);
        return;
      }
      if (s === "\x1b" || (s.startsWith("\x1b[") && s.length < 3)) {
        return;
      }
      // Single char
      if (s.length >= 1) {
        const first = s[0];
        buf = s.slice(1);
        if (first === "w" || first === "k") resolveAndClean("up");
        else if (first === "s" || first === "j") resolveAndClean("down");
        else if (first === "a" || first === "h") resolveAndClean("left");
        else if (first === "d" || first === "l") resolveAndClean("right");
        else resolveAndClean(null);
        return;
      }
    };

    const cleanup = () => {
      process.stdin.removeListener("data", onData);
      if (!wasRaw) process.stdin.setRawMode?.(false);
    };

    process.stdin.on("data", onData);
  });
}

export async function selectFromList<T>(
  items: SelectListItem<T>[],
  prompt: string,
  options?: { initialIndex?: number; quitLabel?: string }
): Promise<T | null> {
  if (items.length === 0) return null;

  const quitLabel = options?.quitLabel ?? "Quit";
  const totalRows = items.length + 1;
  let index = Math.min(Math.max(options?.initialIndex ?? 0, 0), items.length);

  console.log(`\n  ${prompt}\n`);
  while (true) {
    for (let i = 0; i < items.length; i++) {
      console.log((i === index ? "  > " : "    ") + items[i].label);
    }
    console.log((index === items.length ? "  > " : "    ") + quitLabel);

    const key = await waitKey();
    if (key === "quit") return null;
    if (key === "enter") {
      if (index === items.length) return null;
      return items[index].value;
    }
    if (key === "up") index = (index - 1 + totalRows) % totalRows;
    if (key === "down") index = (index + 1) % totalRows;
    process.stdout.write(`\x1b[${totalRows}A\x1b[0J`);
  }
}

function formatRoleLine(role: string, roleConfig: PlanForgeConfig["planner"]): string {
  const extra =
    roleConfig.effort != null
      ? ` (effort: ${roleConfig.effort})`
      : roleConfig.reasoning != null
        ? ` (reasoning: ${roleConfig.reasoning})`
        : "";
  return `  ${role.padEnd(12)}: ${roleConfig.provider.padEnd(6)} / ${roleConfig.model.padEnd(20)}${extra}`;
}

export function printCurrentAiConfig(
  config: Pick<PlanForgeConfig, "planner" | "implementer">,
  heading = "Current AI config"
): void {
  console.log(`\n  ${heading}`);
  console.log(`  ${"-".repeat(heading.length)}`);
  console.log(formatRoleLine("planner", config.planner));
  console.log(formatRoleLine("implementer", config.implementer));
}
